sentinelhub/data_utils.py

```diff
@@ -116,37 +126,46 @@ def statistical_to_dataframe(result_data: List[JsonDict], exclude_stats: Optiona
     return pandas.concat(dfs)
 
 
-def _get_failed_intervals(
-    identifier: str, response_data: List[JsonDict]
-) -> Optional[Dict[str, Union[str, List[Tuple[str, str]]]]]:
-    """Collect failed intervals of a partially failed request
+def _get_failed_intervals(result_data: JsonDict) -> Union[str, List[Tuple[str, str]]]:
+    """Collect failed intervals of a single geometry from the (Batch) Statistical result
 
-    :param identifier: The identifier of the geometry.
-    :param response_data: An input representation of Statistical API response.
-    :return: The identifier of a geometry that has a response status of PARTIAL and the failed intervals.
+    :param result_data: An input representation of the (Batch) Statistical API result of a geometry.
+    :return: The failed intervals for a geometry.
     """
-    failed_intervals = []
-    for interval in response_data:
-        if "error" in interval:
-            failed_intervals.append((interval["interval"]["from"], interval["interval"]["to"]))
-
-    return {"identifier": identifier, "failed_intervals": failed_intervals} if failed_intervals else None
-
-
-def get_failed_statistical_requests(result_data: List[JsonDict]) -> List[Dict[str, Union[str, List[Tuple[str, str]]]]]:
+    if _is_batch_stat(result_data):
+        return (
+            "full time range"
+            if "error" in result_data or not result_data["response"]
+            else [
+                (interval["interval"]["from"], interval["interval"]["to"])
+                for interval in result_data["response"]["data"]
+                if "error" in interval
+            ]
+        )
+
+    return (
+        "full time range"
+        if "error" in result_data
+        else [
+            (interval["interval"]["from"], interval["interval"]["to"])
+            for interval in result_data["data"]
+            if "error" in interval
+        ]
+    )
+
+
+def get_failed_statistical_requests(result_data: List[JsonDict]) -> List[JsonDict]:
     """Collect failed requests of (Batch) Statistical Results
 
     :param result_data: An input representation of (Batch) Statistical API result.
     :return: Failed requests of (Batch) Statistical Results.
     """
-    failed_requests = []
-    for result in result_data:
-        identifier, response = result["identifier"], result["response"]
-        if not response:
-            failed_requests.append({"identifier": identifier})
-        else:
-            failed_intervals = _get_failed_intervals(identifier, response["data"])
-            if failed_intervals:
-                failed_requests.append(failed_intervals)
-
-    return failed_requests
+    if _is_batch_stat(result_data[0]):
```

Comment:
Here you split on whether it is a batch response or a regular one. But in `_get_failed_intervals` you do the same. I suggest you only do it here, and unify `_get_failed_intervals` for both. If you can't, then split it into `_get_failed_intervals_batch` and `_get_failed_intervals_regular` (but i am 95% sure you can do all the 'special' handling here)

Reply:
I'm struggling a bit here. A failed response could be:
- {"id": 1, "identifier": "xxx", "error": {}}
- {"id": 1, "identifier": "xxx", "response": null}
- {"id": 1, "identifier": "xxx", "response": {"data": ["interval": {}, "error": {}]}}
- {"data": ["interval": {}, "error": {}]}
- {"error": {}}

May I have your suggestions on how to build the function for these conditions?

Reply:
so you need a `_get_failed_intervals_response` and then `_get_failed_intervals_batch` can check if it got a response, and if it did it uses `_get_failed_intervals_response`, that should work right?

Reply:
Ah true, will fix this.